README
                              Muldis::Rosetta
                                  README
---------------------------------------------------------------------------

by Darren Duncan <darren@DarrenDuncan.net>

PREFACE

This distribution features the Perl 5 or 6 (depending on which copy of this
README file you are reading) module named Muldis::Rosetta (Full-featured
truly relational DBMS in Perl) and its support structure.

Muldis Rosetta has 2 co-developed versions, which are written in native
Perl 5 and Perl 6 respectively, and which are distributed separately from
each other.  But while the 2 versions would be similar, they are explicitly
not intended to be drop-in substitutable for each other, as each would have
some API differences that optimizes it for its host language.  The actual
Muldis D language specification document will be identical for both,
though.

The Perl 5 version is intended to be the primary version during the present
time, as Perl 5 itself is stable, production-ready, and commonly used now.
Eventually the Perl 6 version is expected to supercede it, once Perl 6
itself has implementations that are production-ready and in common use.

That all being said, development of the Perl 6 version has been temporarily
suspended, to focus resources on the Perl 5 version, and therefore the Perl
6 version is likely out of date by the time you read this.  In the future,
the more-advanced Perl 5 version will be translated to Perl 6 again, either
to re-establish their parallel development, or just to provide snapshots.

Read below for how to get support or keep up to date.

DOCUMENTATION READING ORDER

If you want to read all the Muldis Rosetta documentation that exists now in
this distribution, I recommend doing so in this order:

   1.       README
   2. (opt) Changes
   3.       lib/Muldis/Rosetta.pm
   4.       lib/Muldis/Rosetta/Interface.pm
   5. (opt) lib/Muldis/Rosetta/Validator.pm
   6. (opt) lib/Muldis/Rosetta/Engine/Example.pm
   7.       lib/Muldis/Rosetta/SeeAlso.pod
   8. (opt) TODO

Then you can carry on with anything else.

You may prefer the nicely HTML-formatted versions of the above, which are
available at "http://search.cpan.org/dist/Muldis-Rosetta/".

I also suggest the Muldis-D-Manual distribution, which is available at
"http://search.cpan.org/dist/Muldis-D-Manual/".

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my personal website, "http://www.darrenduncan.net/",
on the page called "Perl Libraries I Made" (name subject to change).  Some
of these modules are also linked to through my business website,
"http://www.muldis.com/", on its front page.

For those of you that are interested in the bleeding edge of development, I
have all parts of the Muldis Rosetta project in one or more public
repositories.  The Perl 5 Muldis Rosetta project is in a Git repository
whose public clone url is "git://github.com/muldis/Muldis-Rosetta.git" and
whose web information url is at "http://github.com/muldis/Muldis-Rosetta".
It is also in a Git repository whose public clone url is
"git://utsl.gen.nz/Muldis-Rosetta" and whose web information url is at
"http://utsl.gen.nz/gitweb/?p=Muldis-Rosetta".  The Perl 6 Muldis Rosetta
project is in a Subversion repository at
"http://svn.pugscode.org/pugs/ext/Muldis-Rosetta/".

You can download slightly older versions of all of my modules from the
public CPAN archives at "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
on my own web server at "http://www.darrenduncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is at
"http://search.cpan.org", and my own DUNCAND author page is at
"http://search.cpan.org/search?author=DUNCAND".

The main CPAN search site url for Muldis Rosetta for Perl 5 is
"http://search.cpan.org/dist/Muldis-Rosetta/"; that is the url you should
link to presently if you want something to consider the "home page" of
Muldis Rosetta; it is best to refer people there by default.

You can download my oldest tarred and gzipped distributions from BACKPAN,
at "http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my
web server, at "http://www.darrenduncan.net/d/perl/archives/", right back
to the libdwg 1.0, first released on 2000 July 23.

SUPPORT

Several public email-based forums exist whose main topic is
the Muldis D language and its implementations, especially
the Muldis Rosetta reference implementation, but also
the Set::Relation module.  They exist so that users of Muldis D or
Muldis Rosetta can help each other, or so that help coming from the
projects' developers can be said once to many people, rather than
necessarily to each individually.  All of these you can reach via
"http://mm.darrenduncan.net/mailman/listinfo".

Alternately, you can purchase more advanced commercial support for various
Muldis D implementations, particularly Muldis Rosetta, from its author by
way of Muldis Data Systems; see "http://www.muldis.com/" for details.

FIN

Share and Enjoy!


Changes
                              Muldis::Rosetta
                                  Changes
---------------------------------------------------------------------------

2011-05-xx

    * Muldis Rosetta version 0.17.0 for Perl 5 is released on CPAN as
    Muldis-Rosetta-0.17.0.tar.gz.  Muldis Rosetta version 0.17.0 for Perl 6
    is not released at all.  The rest of this Changes entry refers only to
    the Perl 5 version, but it also happens that the Perl 6 version was
    also brought fully up to date along with it.
TODO: RESYNC PERL 6 VERSION USING RAKUDO STAR!  SEE TOP TODO FILE ITEM!

    * Updated external dependencies: Test::More to version 0.94,
    namespace::autoclean to version 0.12, Try::Tiny to version 0.09,
    Moose/Class::MOP/etc to version 2.0000, MooseX::Singleton to version
    0.27.

    * Updated the minimum required Perl version to 5.8.3 from 5.8.1; this
    is because Moose 1.09 made the same change, due to its tests failing on
    5.8.1-2 and those versions not being easily installable for testing.

    * (Interface.pm, Example.pm)  Updated all *.pm files that declare
    multiple Perl packages, so that now every single package separately
    declares its (identical) package $VERSION, rather than just the first
    package in each file declaring said, implicitly applied to the others.

    * (Example.pm)  Updated all 3 classes that compose roles so that they
    now specify what version numbers of said roles they require.  Moose had
    added support for declaring/enforcing this in its v1.03 (2010 May 6).

    * (Rosetta.pm, README)  Updated the list of email forums focusing on
    Muldis D and its implementations, first to add a 4th list
    muldis-d-language, and second to update all 4 lists' descriptions.

    * (Rosetta.pm, README, Makefile.PL)  Updated any instances of
    "http://mm.DarrenDuncan.net/mailman/listinfo" so that they are fully
    lower-cased, because the mixed case version doesn't work with some web
    browsers; browsers that don't automatically convert it to lowercase get
    a "There currently are no publically-advertised mailman lists" message.

    * (Example.pm)  In the "INTERFACE" main pod section, fixed old fossil
    where the Muldis D dialects were referred to as "Tiny ..." rather than
    "Standard ...", and also specified one as "Perl 5" rather than "Perl".

    * (SeeAlso.pod)  A few additions to Prospective Extensions, etc.

    * Incremented all copyright year range-ends to 2011.

2010-03-01

    * Muldis Rosetta version 0.16.0 for Perl 5 is released on CPAN as
    Muldis-Rosetta-0.16.0.tar.gz.  Muldis Rosetta version 0.16.0 for Perl 6
    is not released at all.  The rest of this Changes entry refers only to
    the Perl 5 version, and it should not be assumed that the Perl 6
    version incorporates all of said updates, though it may have some of
    them.

    * Oops, the previous release 0.15.0 shipped with code all declared as
    version 0.014000 rather than 0.015000; the metadata was correct though.

    * Updated external dependencies: Try::Tiny to version 0.04, Moose/etc
    to version 0.98, MooseX::Singleton to version 0.22, Class::MOP to
    version 0.98.

    * (Example.pm)  Corresponding to changes introduced in the
    documentation of MooseX::Singleton v0.22:  Updated the new_machine()
    subroutine of the "::Example" module to invoke the "::Machine" class'
    constructor submethod under the name "instance()" rather than "new()",
    because MXS now encourages the use of "instance" over "new", and the
    latter may be deprecated in a future MXS release.

    * (Interface.pm, Example.pm, Validator.pm, SeeAlso.pod)  In
    Interface.pm, updated the Muldis Rosetta public API so that the
    "::Machine"-constructor subroutine which the root module of a Muldis
    Rosetta Engine must provide is now named "select_machine" rather than
    "new_machine", because the new name is more descriptive as to what it
    actually does, being that each "::Machine"-doing class is a singleton.
    In Example.pm, made that change in its "::Example" module.  In
    Validator.pm and SeeAlso.pod, updated any refs to renamed sub's old
    name to use the new name.

    * (Interface.pm, Example.pm, Validator.pm)  Corresponding to changes
    introduced in Muldis D spec v0.96.0:  Updated all 4-element Muldis D
    language names to add the once-again-mandatory 5th (ln_extensions)
    element.  In some documentation, the details of that element are left
    as "...", but in all code or code examples it specifies just the
    mandatory "catalog_abstraction_level" pragma with "rtn_inv_alt_syn".

    * (Interface.pm)  Corresponding to changes introduced in Muldis D spec
    v0.88.0 and v0.108.0:  Update the description of the "trans_nest_level"
    method to omit mentions of "direct boot_stmt transaction management", a
    concept that no longer exists, and also to omit mentions of
    transactions in Muldis D sometimes being tied to "a try/catch context",
    also a concept that no longer exists; now, the *only* way to have a
    transaction start or end in Muldis D code is to for that transaction to
    be attached to the lexical scope of a Muldis D routine, so all
    invocations of that routine are main/nested transactions.

    * (Interface.pm, Validator.pm)  Corresponding to changes introduced in
    Muldis D spec v0.110.0:  Updated all Muldis D language names having a
    ln_extensions element so it specifies the new mandatory
    "op_char_repertoire" pragma with "extended".

    * (SeeAlso.pod)  Other minor updates.

    * Incremented all copyright year range-ends to 2010.

2009-11-09

    * Muldis Rosetta version 0.15.0 for Perl 5 is released on CPAN as
    Muldis-Rosetta-0.15.0.tar.gz.  Muldis Rosetta version 0.15.0 for Perl 6
    is not released at all.  The rest of this Changes entry refers only to
    the Perl 5 version, and it should not be assumed that the Perl 6

Makefile.PL
use 5.008003;
use utf8;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# Make setting optional MakeMaker parameters more readable.
sub OPTIONAL {
   return () unless $ExtUtils::MakeMaker::VERSION ge shift;
   return @_;
}

WriteMakefile(
   NAME => 'Muldis::Rosetta',
   VERSION => '0.16.0',
   ABSTRACT => 'Full-featured truly relational DBMS in Perl',
   AUTHOR => 'Darren Duncan <darren@DarrenDuncan.net>',
   PREREQ_PM => {
      # This list is a copy of "META_ADD:[configure_|build_|]requires"
      # where the other is canonical; this list is for use by MakeMaker
      # itself, which in general wouldn't see the info in the other lists,
      # as the other lists are primarily used when generating the META.yml.
      'Carp::Always' => '0.09',
      'Class::MOP' => '2.0000',
      'ExtUtils::MakeMaker' => '6.48',
      'Moose' => '2.0000',
      'Moose::Role' => '2.0000',
      'MooseX::Singleton' => '0.27',
      'namespace::autoclean' => '0.12',
      'Test::Moose' => '2.0000',
      'Test::More' => '0.94',
      'Try::Tiny' => '0.09',
   },
   OPTIONAL( '6.48',
      MIN_PERL_VERSION => '5.008003',
   ),
   OPTIONAL( '6.31',
      LICENSE => 'lgpl_3',
   ),
   OPTIONAL( '6.46',
      # Use META_ADD instead of META_MERGE so that we can remove
      # any build-time dependencies that MakeMaker will put into
      # the requires field.  (Comment From DBD::SQLite; do *we* need that?)
      META_ADD => {
         configure_requires => {
            'ExtUtils::MakeMaker' => '6.48',
         },
         build_requires => {
            'ExtUtils::MakeMaker' => '6.48',
            # These are bundled with Perls 5.12.0+ but not earlier Perls.
            'Test::More' => '0.94',
            # These are not bundled with any version of Perl.
            'Carp::Always' => '0.09',
         },
         requires => {
            # ALWAYS REQUIRES:

            'perl' => '5.008003', # as MIN_PERL_VERSION no affect META.yml
            # These are not bundled with any version of Perl.
            'Class::MOP' => '2.0000',
            'Moose' => '2.0000',
            'Moose::Role' => '2.0000',
            'MooseX::Singleton' => '0.27',
            'namespace::autoclean' => '0.12',
            'Try::Tiny' => '0.09',

            # VALIDATOR MODULE REQUIRES:

            # These are bundled with Perls 5.12.0+ but not earlier Perls.
            'Test::More' => '0.94',
            # These are not bundled with any version of Perl.
            'Test::Moose' => '2.0000',
         },
         resources => {
            homepage => 'http://www.muldis.com/',
            license => 'http://www.fsf.org/licensing/licenses/lgpl.html',
            MailingList => 'http://mm.darrenduncan.net/mailman/listinfo',
            repository => 'http://github.com/muldis/Muldis-Rosetta',
         },
      },
   ),
);

1;
